infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java

```diff
@@ -0,0 +1,590 @@
+import org.corfudb.infrastructure.management.PollReport;
+import org.corfudb.protocols.wireprotocol.NodeState;
+import org.corfudb.protocols.wireprotocol.SequencerMetrics.SequencerStatus;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.QuorumUnreachableException;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+import org.corfudb.runtime.view.Layout;
+import org.corfudb.runtime.view.QuorumFuturesFactory;
+import org.corfudb.util.NodeLocator;
+import org.corfudb.util.concurrent.SingletonResource;
+
+/**
+ * Remote Monitoring Service constitutes of failure and healing detection and handling.
+ * This service is responsible for heartbeat and aggregating the cluster view. This is
+ * updated in the shared context with the management server which serves the heartbeat responses.
+ * The failure detector updates the unreachable nodes in the layout.
+ * The healing detector heals nodes which were previously marked unresponsive but have now healed.
+ * Created by zlokhandwala on 11/2/18.
+ */
+@Slf4j
+public class RemoteMonitoringService implements IService {
+
+    /**
+     * Detectors to be used to detect failures and healing.
+     */
+    @Getter
+    private IDetector failureDetector;
+    @Getter
+    private IDetector healingDetector;
+
+    /**
+     * Detection Task Scheduler Service
+     * This service schedules the following tasks every POLICY_EXECUTE_INTERVAL (1 sec):
+     * - Detection of failed nodes.
+     * - Detection of healed nodes.
+     */
+    @Getter
+    private final ScheduledExecutorService detectionTasksScheduler;
+    /**
+     * To dispatch tasks for failure or healed nodes detection.
+     */
+    @Getter
+    private final ExecutorService detectionTaskWorkers;
+
+    private final ServerContext serverContext;
+    private final SingletonResource<CorfuRuntime> runtimeSingletonResource;
+
+    /**
+     * Cluster state context to update the connectivity graph populated by the failure and healing
+     * detectors.
+     */
+    private final ClusterStateContext clusterStateContext;
+
+    /**
+     * Interval in executing the failure detection policy.
+     * In milliseconds.
+     */
+    @Getter
+    private static final long POLICY_EXECUTE_INTERVAL = 1000;
+
+    /**
+     * Future for periodic failure and healed nodes detection task.
+     */
+    private Future failureDetectorFuture = null;
+    private Future healingDetectorFuture = null;
+
+    /**
+     * The management agent attempts to bootstrap a NOT_READY sequencer if the
+     * sequencerNotReadyCounter counter exceeds this value.
+     */
+    private static final int SEQUENCER_NOT_READY_THRESHOLD = 3;
+
+    /**
+     * This tuple maintains, in an epoch, how many heartbeats the primary sequencer has responded
+     * in not bootstrapped (NOT_READY) state.
+     */
+    @Getter
+    @Setter
+    @AllArgsConstructor
+    private class SequencerNotReadyCounter {
```

Comment:
I'm not sure if this class carries its weight or adds more clarification to code. With the current usage maybe using variables would be enough.

Reply:
Maybe an immutable form of more generic class (e.g. tuple of epic and counter) can be used instead.